### DISCUSSION TITLE
Peculiar types for `Path`s that differ between Scala 2 and 3

### DISCUSSION THREAD

**Discussion by SethTisue** [lihaoyi/os-lib#153] [General]
starting at `os.root`, what types result as we add path segments one at a time?

in Scala 2.13.10, we start at `Path`, take a brief detour to `os.root.ThisType`, and then we're back at `Path`

```scala
scala> os.root
val res4: os.Path = /

scala> os.root / "a"
val res5: os.root.ThisType = /a

scala> os.root / "a" / "b"
val res6: os.Path = /a/b

scala> os.root / "a" / "b" / "c"
val res7: os.Path = /a/b/c
```

on Scala 3, the types just keep getting longer:

```scala
scala> os.root
val res0: os.Path = /
                                                                                                    
scala> os.root / "a"
val res1: os.root.ThisType = /a
                                                                                                    
scala> os.root / "a" / "b"
val res2: os.root.ThisType#ThisType = /a/b
                                                                                                    
scala> os.root / "a" / "b" / "c"
val res3: os.root.ThisType#ThisType#ThisType = /a/b/c
```

not sure whether we could do better here, either with changes to the library, or with changes to the compiler, or both

**Comment by lefou** [comment]
I think Scala 3 is more correct here, as we use a `type ThisType <: BasePath` as return value. Maybe the use of `this.type` helps here.

**Reply by lefou** [reply]
No, I think `this.type` can't work, as we want to return new instances of immutable classes.

**Comment by lefou** [comment]
* I opened PR #154

**Reply by SethTisue** [reply]
Nice! Glad it turned out to be tractable to do something about this.

**Reply by SethTisue** [reply]
incidentally the context where this came up for me was https://github.com/scalacenter/docs.scala-lang/pull/21
